eventuate-crdt/src/main/scala/com/rbmhtechnology/eventuate/crdt/RGArray.scala

```diff
@@ -0,0 +1,165 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.rbmhtechnology.eventuate.crdt
+
+import akka.actor._
+import com.rbmhtechnology.eventuate._
+
+import scala.annotation.tailrec
+import scala.concurrent.Future
+import scala.collection.immutable.Vector
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * Operation-based Replicated Growable Array CRDT.
+ *
+ * @tparam A Entry value type.
+ *
+ * @see [[http://hal.upmc.fr/docs/00/55/55/88/PDF/techreport.pdf A comprehensive study of Convergent and Commutative Replicated Data Types]], specification 15
+ */
+case class RGArray[A](vertices: Vector[Vertex[A]] = Vector.empty[Vertex[A]], lastPos: Int = 0) extends CRDTFormat {
+  def value: Vector[A] = vertices.filter(_.tombstone.isEmpty).map(_.value)
+
+  def getVertex(index: Int): Vertex[A] = vertices.iterator.filter(_.tombstone.isEmpty).drop(index).next
+
+  def insertRight(value: A, newPos: Position): RGArray[A] = insertRight(Position.head, value, newPos)
+
+  def insertRight(index: Position, value: A, id: Position): RGArray[A] = {
+    val newVertex = Vertex(value, id)
+    val nextPos = math.max(id.order, lastPos)
+    if (vertices.isEmpty) {
+      RGArray(vertices :+ newVertex, nextPos)
+    } else {
+      val found = if (index == Position.head) 0 else vertices.indexWhere((v: Vertex[A]) => v.pos == index) + 1
+      val idx: Int = skipOver(id, found)
+      RGArray((vertices.take(idx) :+ newVertex) ++: vertices.drop(idx), nextPos)
+    }
+  }
+
+  /**
+   * Returns a new RGArray with vertex identified by provided `index` marked as tombstoned at provided `time`.
+   *
+   * @param index An absolute position of an element to remove.
+   * @param time A timestamp at which removal has happened.
+   */
+  def delete(index: Position, time: VectorTime) = {
+    this.copy(vertices = vertices.map((v: Vertex[A]) => if (v.pos == index) v.copy(tombstone = Some(time)) else v))
+  }
+
+  /**
+   * Prunes all removed values, that have been tombstoned in `timestamp`s casual past.
+   *
+   * @param timestamp Last stable timestamp, reached by all corresponding replicas.
+   * @return A new RGArray without tombstoned values, that had happened before `timestamp`.
+   */
+  def prune(timestamp: VectorTime) = {
+    val pruned = vertices.filter(x => {
+      x.tombstone match {
+        case Some(time) if time < timestamp => false
+        case _                              => true
+      }
+    })
+    this.copy(vertices = pruned)
+  }
+
+  @tailrec
+  private def skipOver(pos: Position, idx: Int): Int = {
+    if (idx < vertices.length && vertices(idx).pos > pos) skipOver(pos, idx + 1) else idx
+  }
+}
```

Comment:
Please review parameter naming in `RGArray` class:

- parameters with `pos` in name have type `Position` (expected) or `Int` (not expected)
- parameters of type `Position` have names containing `pos` (expected), `id` (not expected) or `index` (not expected)
- ...

I skip listing other examples, hope it's clear what I mean.